File: solutions/C++/permutations_swap.cpp
#include <vector>
using namespace std;

class Solution {
public:
    /**
     * Generate all permutations using backtracking with swapping approach.
     * Time: O(n! * n), Space: O(n!) for result
     */
    vector<vector<int>> permute(vector<int> nums) {
        vector<vector<int>> result;
        backtrack(nums, 0, result);
        return result;
    }

private:
    void backtrack(vector<int>& nums, int first, vector<vector<int>>& result) {
        // Base case: all elements are placed
        if (first == nums.size()) {
            result.push_back(nums);
            return;
        }

        for (int i = first; i < nums.size(); i++) {
            // Swap
            swap(nums[first], nums[i]);
            // Backtrack
            backtrack(nums, first + 1, result);
            // Swap back
            swap(nums[first], nums[i]);
        }
    }
};

// Test
int main() {
    Solution sol;
    vector<vector<int>> result = sol.permute({1, 2, 3});
    // Output: [[1,2,3], [1,3,2], [2,1,3], [2,3,1], [3,1,2], [3,2,1]]
    return 0;
}

File: solutions/Rust/permutations_swap.rs
pub fn permutations_swap(nums: Vec<i32>) -> Vec<Vec<i32>> {
    let mut result = vec![];
    let mut nums = nums;
    backtrack(&mut nums, 0, &mut result);
    result
}

fn backtrack(nums: &mut Vec<i32>, first: usize, result: &mut Vec<Vec<i32>>) {
    // Base case: all elements are placed
    if first == nums.len() {
        result.push(nums.clone());
        return;
    }

    for i in first..nums.len() {
        // Swap
        nums.swap(first, i);
        // Backtrack
        backtrack(nums, first + 1, result);
        // Swap back
        nums.swap(first, i);
    }
}

fn main() {
    let result = permutations_swap(vec![1, 2, 3]);
    println!("{:?}", result);
    // Output: [[1, 2, 3], [1, 3, 2], [2, 1, 3], [2, 3, 1], [3, 1, 2], [3, 2, 1]]
}
